### PR TITLE
NetworkDevice Implementation and miscellaneous changes

### PR DESCRIPTION
**Pull request by blobjim**
I created an alternative to the static Client/Server/Parse system. It uses UDP only and allows communication between two IP addresses (although I can pretty easily expand it to allow any number of connections). I added a class to the Communication.Constants namespace, MessageTypeID, for use with the NetworkDevice class. I also added/improved some methods in the Utilities.UtilData class. All UtilData methods should work as is, but are now more flexible. The previous string serialization methods have been removed and replaced with methods that simply use ones from .NET with UTF8 encoding. I added example usage of the NetworkDevice class to the TestSuit 'Main' class.

### CODE REVIEW

**Comment by CaiB** [comment]
I'll have to look through this in more detail, but two issues I see right away, there appear to be stray changes to `IO/BeagleBone/BBBPinManager.cs`, could you revert these?

Also, could you please split the changes made in `TestSuite/Main.cs` to better fit in with the other test functions? Currently you're doing networking tasks on every run of the TestSuite project, which isn't the goal of that project.

Since we were never informed of this being done or the reasoning behind this, could you explain how this new system is better than our current Client/Server and TCP/UDP system?

**Comment by Baldstrom** [comment]
If we decide to merge this, I think this will need much more testing than what is currently described in the `TestSuite`, have you done any other tests? For example remote testing? If so, can you describe the tests you have run?

This commit will take quite a while to review fully as it is relatively large and a major addition to communications. A lot of the review will have to do with code style too, and from my first scan through, the biggest violation you have is probably (in multiple places) where you have an if statement with no curly braces, which are required for every if statement per our style guide. Otherwise style looks pretty good, but we will go into more depth later.

Another big thing to note is that if this code is merged, you will be responsible for the corresponding documentation in OneNote; if you can have samples of that prepared before we merge that would be even better and help us review your pull request.

**Comment by blobjim** [comment]
I'll work on adding unit testing, updating the code style, and writing more formal documentation. I'm also going to make some other internal changes to NetworkDevice like adding a more reliable way for a NetworkDevice to connect to another (internal change only) and using a Socket instead of UdpClient if I can.

**Comment by Baldstrom** [comment]
Okay, cool. The other major thing is that we are already planning to do a whole networking re-write per our documentation (`Roadmap`, also newly `Networking Over-Haul Ideas`), so I am not sure what we will decide to do with this code. When we do that over-haul, either this will already be merged and later marked deprecated or completely removed if it conflicts with the overhaul somehow. If we wait on this pull request until the networking over-haul, we can use the ideas here to shape the features, I suppose.

Also, could you comment on how this new system is better than our current Client/Server and TCP/UDP system? Cai asked about that earlier and I think it's important to note for how we decide to proceed with it.

**Comment by blobjim** [comment]
The NetworkDevice class is mostly ported from a game I've been writing on and off in Java, with a few changes to make it feel more like the system you guys created (registering parsers, packet date/time). It uses UDP only to be a bit simpler than a mix of UDP/TCP and since TCP is stream-oriented using it to pass messages isn't optimal (like having to set TCP_NODELAY or something). The only downside I see to using a custom reliable-message system over UDP is that TCP is more optimized for latency and other stuff. I found that one receiving/processing thread works pretty well, and using more threads just complicates things (although using a thread pool for processing would probably work better, but possibly lead to multithreading problems). The class is obviously object oriented, which helps a lot with organization, code reuse, and being able to test code within a single process. I don't know whether the rover will have one or multiple IP addresses so I didn't implement multiple clients (although my Java implementation has them). I assume the base station will have to initiate the connection to the rover so the base station will act as a 'client' in some ways while also being authoritative.

Basically, if you decide to overhaul the current Client/Server system, consider the way I've implemented this class. :)

**Comment by blobjim** [comment]
Here's a link to a tiny bit of documentation: https://1drv.ms/f/s!AsUeiBkG_NEihRDoE5upiJBLzfik

**Comment by blobjim** [comment]
Everything should work properly now. :3
